### PR TITLE
feat: add telemetry and e2e tests for Duplicate Contact Logic

### PR DESCRIPTION
**Pull request by jkuester**
# Description

@ChinHairSaintClair in lieu of a final round of review on https://github.com/medic/cht-core/pull/9609, I wanted to just open this PR into your branch with my suggested changes:

- Add telemetry recording to collect data on performance, duplicates found, and the times a user overrode the dupe checking
- Made a few behavioral tweaks to the panels (explained in the comments below).
- Added e2e tests for everything.

Please take some time to review these code changes and let me know if you want to do anything differently or if you can think of anything else that I missed.  Once you are good with the code/tests here, just merge this PR into your branch.  Then, we should be ready to merge your PR to `medic/cht-core/master`!

The good news is that this branch [is building successfully in CI](https://github.com/medic/cht-core/actions/runs/14277072123)!

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/angular.json [hunk 1]
```diff
@@ -81,7 +81,7 @@
                 {
                   "type": "initial",
                   "maximumWarning": "2mb",
-                  "maximumError": "5mb"
+                  "maximumError": "6mb"
```
Comment (jkuester) [inline webapp/angular.json:84]: I did some double-checking.  We have been hovering ~4.99mb since at least `4.10` and this finally pushed us over.  

In general we definitely want to avoid anything that significantly increases the bundle size, but it is not really possible to add logic without slightly increasing the size.  So, bumping this another mb to give us more breathing room...

File: webapp/src/ts/components/duplicate-contacts/duplicate-contacts.component.html [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
-    <p *ngIf="error" class="error_msg"><strong>{{ error!.toString() }}</strong></p>
+    <p *ngIf="error" class="loading-status"><strong>{{ error }}</strong></p>
     <mm-contact-summary-content
-      [fields]="_summary?.fields"
+      [fields]="contactSummary?.fields"
       [contactsLoadingSummary]="isLoading">
     </mm-contact-summary-content>
   </div>
   <mat-action-row>
-    <button class="btn submit btn-primary" *ngIf="error" (click)="_loadContactSummary()">{{ 'duplicate_check.contact.retry' | translate }}</button>
```
Comment (jkuester) [inline webapp/src/ts/components/duplicate-contacts/duplicate-contacts.component.html:17]: Happy to discuss this further, but after thinking about this more, I did not really see the point of having the "Retry" button.  The contact-summary calculation is not really the kind of thing that we expect to fail sometimes, but may work if you try it again. If it fails once, it will likely continue to fail.

So, instead of a giving a Retry button, I just made the "Open" button always visible. If the contact-summary fails, the user will see a message indicating it failed to load, but they can still use the "Open" button to navigate to the contact if they want.

File: webapp/src/ts/components/duplicate-contacts/duplicate-contacts.component.ts [hunk 3]
```diff
@@ -33,44 +33,30 @@ import { Contact } from '@medic/cht-datasource';
...
   }
   
-  _loadContactSummary = async () => {
-    if (!this.duplicate?._id){
-      return;
-    }
-    
-    if (this.isLoading){
+  loadContactSummary = async () => {
+    if (this.contactSummary || !this.duplicate?._id || this.isLoading){
```
Comment (jkuester) [inline webapp/src/ts/components/duplicate-contacts/duplicate-contacts.component.ts:52]: I added `this.contactSummary` to the check here so we don't re-load the contact summary once it has been calculated (e.g. if the user just opens and closes the panel repeatedly).

File: webapp/src/ts/components/duplicate-contacts/duplicate-contacts.component.ts [hunk 4]
```diff
@@ -33,44 +33,30 @@ import { Contact } from '@medic/cht-datasource';
...
       return;
     }
 
     this.isLoading = true;
-    this.error = undefined;
     
     try {
-      const sanitizedContact = Object.keys(this.duplicate).reduce((acc, key) => {
```
Comment (jkuester) [inline webapp/src/ts/components/duplicate-contacts/duplicate-contacts.component.ts:66]: See [my comment](https://github.com/medic/cht-core/pull/9609#discussion_r2012117784) on the original PR regarding avoiding this sanitation.  Happy to continue discussing here!

File: webapp/src/css/enketo/medic.less [hunk 5]
```diff
@@ -445,6 +445,12 @@
   }
 
   #duplicate_contacts {
+    display: none;
```
Comment (jkuester) [inline webapp/src/css/enketo/medic.less]: I think we only want to show the duplicates panel _on the last page_ of the form (right before the "Submit" button).  If a user tries to submit a form and there are dupes found, then they use the "Previous" button to go back to an earlier page in the form, they should not still see the duplicates panel.

The easiest way I could come up with to only show the duplicates on the last page was to put it into the `.form-footer` and then control the display here based on if the `.form-footer` had the `.end` class.  All the logic for adding the `.end` class to the `.form-footer` is already existing (and controls the "Submit" vs "Next" button).

File: webapp/src/css/enketo/medic.less [hunk 6]
```diff
@@ -472,8 +478,8 @@
         font-size: @font-medium;
         justify-content: center;
         display: flex;
-        .error_msg {
-          color: #a94442;
+        .loading-status {
```
Comment (jkuester) [inline webapp/src/css/enketo/medic.less:476]: Matches other places in the app where the loading status is greyed out (instead of being red).

File: webapp/src/ts/modules/contacts/contacts-edit.component.html [hunk 7]
```diff
@@ -10,23 +10,20 @@
   <div class="col-sm-8 item-content material" [hidden]="loadingContent || contentError">
     <div class="card">
       <mm-enketo formId="contact-form" [editing]="enketoContact?.docId" [status]="enketoStatus" (onSubmit)="save()" (onCancel)="navigationCancel()">
-        <div duplicate-contacts id="duplicate_contacts" [ngStyle]="{ display: duplicates.length > 0 ? 'block' : 'none' }" class="question invalid-value">
-          <p class="results_header invalid-value-msg active">{{ duplicateMessage }}</p>
-          <div class="divider"></div>
+        <div duplicate-contacts id="duplicate_contacts" *ngIf="duplicates.length" [ngClass]="{'invalid-value': enketoError}" class="question">
+          <p [ngClass]="{'invalid-value-msg': enketoError}" class="results_header active">{{ duplicateMessage }}</p>
```
Comment (jkuester) [inline webapp/src/ts/modules/contacts/contacts-edit.component.html:14]: Varying the color here based on the presence of `enketoError`.  The idea is that when the user tries to submit the form and dupes are found, the dupes should be highlighted in red.  However, if the user then checks the "Acknowledge" button, the error should be cleared along with the red error highlighting (though we still leave the duplicate list for reference).  

My hope is that this gives a bit of a visual queue to the user that they should try submitting again.

File: webapp/src/ts/modules/contacts/contacts-edit.component.ts [hunk 8]
```diff
@@ -298,6 +302,9 @@ export class ContactsEditComponent implements OnInit, OnDestroy, AfterViewInit {
     if (this.enketoError) {
       this.globalActions.setEnketoError(null);
     }
+
+    this.duplicatesAcknowledged = false;
+    this.duplicates = [];
```
Comment (jkuester) [inline webapp/src/ts/modules/contacts/contacts-edit.component.ts]: This `resetFormError` will get called any time the user changes something in the form.  So, if they try to "Submit" and get a dupes error, then they go and change the `name` of the new contact they are creating, we want to clear out the duplicate state here so that we can come at it with a clean new check.

File: webapp/src/ts/modules/contacts/contacts-edit.component.ts [hunk 9]
```diff
@@ -357,6 +364,12 @@ export class ContactsEditComponent implements OnInit, OnDestroy, AfterViewInit {
         // Updating fields before save. Ref: #6670.
         $('form.or').trigger('beforesave');
 
+        if (this.duplicatesAcknowledged && this.duplicates.length) {
+          this.telemetryService.record(
+            ['enketo', 'contacts', this.enketoContact.type, 'duplicates_acknowledged'].join(':')
+          );
```
Comment (jkuester) [inline webapp/src/ts/modules/contacts/contacts-edit.component.ts:367]: Record each time a user acknowledges the duplicates.

File: webapp/src/ts/services/deduplicate.service.ts [hunk 10]
```diff
@@ -29,22 +31,27 @@ export class DeduplicateService {
...
       });
-    }).sort((a: Contact.v1.Contact, b: Contact.v1.Contact) => {
-      // Desc order - reverse order by switching props
-      return new Date(b.reported_date || 0).getTime() - new Date(a.reported_date || 0).getTime();
-    });
+
+    this.telemetryService.record(
+      ['enketo', 'contacts', contactType, 'duplicates_found'].join(':'),
+      duplicates.length
+    );
```
Comment (jkuester) [inline webapp/src/ts/services/deduplicate.service.ts:53]: Record how many duplicates were found each time this check is run.

File: webapp/src/ts/services/form.service.ts [hunk 11]
```diff
@@ -337,22 +339,30 @@ export class FormService {
...
-    ) : [];
+    const perfTracking = this.performanceService.track();
+
+    try {
+      const siblings = await this.contactsService.getSiblings(doc);
+      return this.deduplicateService.getDuplicates(doc, contactType, siblings, duplicateCheck);
+    } finally {
+      perfTracking?.stop({
+        name: ['enketo', 'contacts', contactType, 'duplicate_check'].join(':')
+      });
```
Comment (jkuester) [inline webapp/src/ts/services/form.service.ts:358]: Measure how long the whole dupe-checking process took (including querying the siblings and running the dupe check logic).

**Comment by ChinHairSaintClair** [comment]
Thank you for your thoughtful feedback and all the effort you've put in to get this feature into the CHT, we really appreciate it @jkuester.

With regards to your comments on 'only showing the panel on the last page' and 'clearing duplicate state after any form change':

We understand the intent might be to reduce cognitive load, reduce screen real-estate usage, earlier in the form. However, one of our primary design goals is to enable "side-by-side" comparison of field values. If the user receives a dupe warning and wants to validate or update individual fields, restricting the duplicates to only the final screen makes that difficult - they'd have to memorize or flip back and forth.

We're aiming for a smoother experience where the users can glance at the dupe info while editing relevant fields. Keeping the panel visible - even on earlier pages after detection - is a deliberate choice to support that. Perhaps there's something we're not considering?

We agree that stale state can be misleading, but in our case, clearing duplicates on any change might be a little too aggressive. Since we're not doing real-time dedupe checking, we risk erasing helpful context while the user is still verifying or resolving an issue. Perhaps a more 'real-time check' is something we can look into after MVP?

**Comment by jkuester** [comment]
> We understand the intent might be to reduce cognitive load, reduce screen real-estate usage, earlier in the form. However, one of our primary design goals is to enable "side-by-side" comparison of field values. If the user receives a dupe warning and wants to validate or update individual fields, restricting the duplicates to only the final screen makes that difficult - they'd have to memorize or flip back and forth.
> 
> We're aiming for a smoother experience where the users can glance at the dupe info while editing relevant fields. Keeping the panel visible - even on earlier pages after detection - is a deliberate choice to support that. Perhaps there's something we're not considering?

@ChinHairSaintClair great push-back here!  I think you did an excellent job describing the tradeoffs and it is really helpful to know this is intended/desired behavior.  With all this in mind, I am good with leaving the duplicates visible both on previous pages and after a value in the form has changed.  :+1: (and have just pushed some changes to update this behavior)  I did leave in place the logic to clear the red background from the duplicates when a value changes in the form (based on the presence of `enketoError`), but let me know if you think that is unhelpful/confusion.

**Comment by ChinHairSaintClair** [comment]
Thank you @jkuester! I really like that subtle nudge to re-evaluate the form. Happy to see these changes in the base branch🙂!

**Comment by ChinHairSaintClair** [comment]
@jkuester we have 3 failing checks, two of which seem to be timeouts, should these be ignored?

`ci-webdriver-default`
should work with incorrect locale 
- timeout

`ci-webdriver-default-workflows`
the LHS button should show the appropriate label when multiple places exist at the current level
- timeout

`ci-webdriver-default-mobile-all`
should redirect to login page when session is expired
- at async Object.goToUrl (/home/runner/work/cht-core/cht-core/tests/page-objects/default/common/common.wdio.page.js:325:5)
- at async Object.goToAdminUpgrade (/home/runner/work/cht-core/cht-core/tests/page-objects/default/users/user.wdio.page.js:35:3)
- at async Context.<anonymous> (/home/runner/work/cht-core/cht-core/tests/e2e/default/navigation/bfcache.wdio-spec.js:51:7)

**Comment by jkuester** [comment]
Yeah, the good news is that we have a lot of automated e2e tests validating workflows. The bad news is that test flakiness is a constant battle :sweat:.  I don't think these failures are related to our code changes in any way (any they will probably resolve themselves if you re-run the jobs once or twice.  

The same CI jobs have already for the same commit over on the medic repo: https://github.com/medic/cht-core/actions/runs/14311610099

So, really there is no need to get these to pass before merging back to the original PR. :+1:
